extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryFrameProcessor.java

```diff
@@ -0,0 +1,550 @@
+  {
+    if (frameWriter == null) {
+      final ColumnSelectorFactoryMaker csfm = ColumnSelectorFactoryMaker.fromRAC(rac);
+      final ColumnSelectorFactory frameWriterColumnSelectorFactory = makeVirtualColumnsForFrameWriter(
+          partitionBoostVirtualColumn,
+          jsonMapper,
+          query
+      ).wrap(csfm.make(rowId));
+      frameWriter = frameWriterFactory.newFrameWriter(frameWriterColumnSelectorFactory);
+      currentAllocatorCapacity = frameWriterFactory.allocatorCapacity();
+    }
+  }
+
+  /**
+   * @param rac   {@link RowsAndColumns} to be written to frame
+   * @param rowId Counter to keep track of how many rows are added
+   * @throws IOException
+   */
+  public void writeRacToFrame(RowsAndColumns rac, AtomicInteger rowId) throws IOException
+  {
+    final int numRows = rac.numRows();
+    rowId.set(0);
+    while (rowId.get() < numRows) {
+      final boolean didAddToFrame = frameWriter.addSelection();
+      if (didAddToFrame) {
+        incrementBoostColumn();
+        rowId.incrementAndGet();
+      } else if (frameWriter.getNumRows() == 0) {
+        throw new FrameRowTooLargeException(currentAllocatorCapacity);
+      } else {
+        flushFrameWriter();
+        return;
+      }
+    }
+    flushFrameWriter();
+  }
+
+  @Override
+  public void cleanup() throws IOException
+  {
+    FrameProcessors.closeAll(inputChannels(), outputChannels(), frameWriter);
+  }
+
+  /**
+   * @return Number of rows flushed to the output channel
+   * @throws IOException
+   */
+  private long flushFrameWriter() throws IOException
+  {
+    if (frameWriter == null || frameWriter.getNumRows() <= 0) {
+      if (frameWriter != null) {
+        frameWriter.close();
+        frameWriter = null;
+      }
+      return 0;
+    } else {
+      final Frame frame = Frame.wrap(frameWriter.toByteArray());
+      Iterables.getOnlyElement(outputChannels()).write(new FrameWithPartition(frame, FrameWithPartition.NO_PARTITION));
+      frameWriter.close();
+      frameWriter = null;
+      return frame.numRows();
+    }
+  }
+
+  /**
+   * @param frame Row based frame to be converted to a {@link RowsAndColumns} object
+   * Throw an exception if the resultant rac used goes above the guardrail value
+   */
+  private void convertRowFrameToRowsAndColumns(Frame frame)
+  {
+    final RowSignature signature = frameReader.signature();
+    RowBasedFrameRowsAndColumns frameRowsAndColumns = new RowBasedFrameRowsAndColumns(frame, signature);
+    LazilyDecoratedRowsAndColumns ldrc = new LazilyDecoratedRowsAndColumns(
+        frameRowsAndColumns,
+        null,
+        null,
+        null,
+        OffsetLimit.limit(Integer.MAX_VALUE),
+        null,
+        null
+    );
+    frameRowsAndCols.add(ldrc);
+    if (frameRowsAndColumns.numRows() > maxRowsMaterialized) {
+      throw new MSQException(new TooManyRowsInAWindowFault(frameRowsAndColumns.numRows(), maxRowsMaterialized));
+    }
```

Comment:
frame ~ 1MB or 8MBs. When would this exception be hit, given that we know that frameRowsAndColumns.numRows() == frame.numRows()? Also, the size occupied on the heap would not be more than 8MBs right?

Reply:
Yes, the failures are handled only at the rows and col level and not at the frame level, the code is updated